### PR TITLE
oast: Support interactsh

### PR DESCRIPTION
**Pull request by ricekot**
image

### CODE REVIEW

**Comment by ricekot** [comment]
This PR is now ready for a review. Not sure what to do about the LGTM alert, as that algorithm is used by Interactsh.

**Review by psiinon — APPROVED** [review]
Not actually tried it, but the code looks sane :)

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/services/boast/BoastService.java [hunk 1]
```diff
@@ -40,7 +38,7 @@
 
     private List<BoastServer> registeredServers = new ArrayList<>();
     private final ScheduledExecutorService executorService =
-            Executors.newSingleThreadScheduledExecutor(new BoastThreadFactory("ZAP-OAST-BOAST-"));
+            Executors.newSingleThreadScheduledExecutor(new OastThreadFactory("ZAP-OAST-BOAST-"));
```
Comment (kingthorin) [inline addOns/oast/src/main/java/org/zaproxy/addon/oast/services/boast/BoastService.java:41]: Tweak the thread name string too?

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/services/interactsh/InteractshEvent.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                new HttpMessage(
+                        new HttpRequestHeader(),
+                        new HttpRequestBody(),
+                        new HttpResponseHeader(),
+                        new HttpResponseBody());
+        msg.setTimeSentMillis(timestamp.toEpochMilli());
+        if ("http".equals(protocol)) {
+            extractAndSetMsg(msg.getRequestHeader(), msg.getRequestBody(), rawRequest);
+            extractAndSetMsg(msg.getResponseHeader(), msg.getResponseBody(), rawResponse);
+        } else if ("dns".equals(protocol)) {
```
Comment (kingthorin) [inline addOns/oast/src/main/java/org/zaproxy/addon/oast/services/interactsh/InteractshEvent.java:99]: Do these conditions need to ignore case?

Comment (kingthorin) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/services/interactsh/InteractshEvent.java:99]: More below too.

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/services/interactsh/InteractshOptionsPanelTab.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            authToken = new ZapTextField();
+        }
+        return authToken;
+    }
+
+    ZapNumberSpinner getPollingFrequencySpinner() {
+        if (pollingFrequencySpinner == null) {
+            pollingFrequencySpinner =
+                    new ZapNumberSpinner(
+                            InteractshParam.MINIMUM_POLLING_FREQUENCY, 60, Integer.MAX_VALUE);
```
Comment (kingthorin) [inline addOns/oast/src/main/java/org/zaproxy/addon/oast/services/interactsh/InteractshOptionsPanelTab.java:120]: I'm not sure if this is seconds or minutes, but max int seconds is like 68 years. I know there are arguments for not limiting people but....

File: addOns/oast/oast.gradle.kts [hunk 4]
```diff
@@ -40,5 +40,7 @@ crowdin {
 dependencies {
     compileOnly(parent!!.childProjects["graaljs"]!!)
 
+    implementation("com.github.0xshamil:java-xid:1.0.0")
```
[the remark this answers is not in the document]
Comment (ricekot) [reply addOns/oast/oast.gradle.kts:43]: XIDs are used by the interactsh CLI client as well: https://github.com/projectdiscovery/interactsh, which is why I ended up using them here too. The length of the string has to be 20 characters, so I probably cannot use UUIDs. I'll try with a random string instead.

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/services/boast/BoastService.java (hunk 1 above)
Comment (ricekot) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/services/boast/BoastService.java:41]: This thread is only going to poll the registered BOAST servers, hence it's still BOAST. The Interactsh thread contains Interactsh in the name: [link](https://github.com/zaproxy/zap-extensions/pull/3120/files#diff-b5dc454da8152981512ff7179b4f6bdf35ebdb69d6289f23471d781b0b07b9d8R81).

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/services/interactsh/InteractshEvent.java (hunk 2 above)
Comment (ricekot) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/services/interactsh/InteractshEvent.java:99]: Nope, this is how they are always sent, e.g. see [this link](https://github.com/projectdiscovery/interactsh/blob/f5b53a7b8be329ee6d4936ed753e3c46472e5726/pkg/server/http_server.go#L100) for "http".

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/services/interactsh/InteractshOptionsPanelTab.java (hunk 3 above)
Comment (ricekot) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/services/interactsh/InteractshOptionsPanelTab.java:120]: Agreed, we can have an upper limit... what do you suggest?

Comment (kingthorin) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/services/interactsh/InteractshOptionsPanelTab.java:120]: I dunno an hour, 6 hours. Six even seems like a stretch to me, but I think that should be a ton of flexibility.

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/services/boast/BoastService.java (hunk 1 above)
Comment (kingthorin) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/services/boast/BoastService.java:41]: Okay, good.

File: addOns/oast/oast.gradle.kts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kingthorin) [reply addOns/oast/oast.gradle.kts:43]: Maybe: https://docs.oracle.com/javase/8/docs/api/javax/transaction/xa/Xid.html

[the remark this answers is not in the document]
Comment (ricekot) [reply addOns/oast/oast.gradle.kts]: A random alphanumeric string works :+1:.

**Comment by ricekot** [comment]
Tweaked.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/services/boast/BoastOptionsPanelTab.java [hunk 5]
```diff
@@ -94,8 +94,7 @@ ZapTextField getBoastUri() {
     ZapNumberSpinner getPollingFrequencySpinner() {
         if (pollingFrequencySpinner == null) {
             pollingFrequencySpinner =
-                    new ZapNumberSpinner(
-                            BoastParam.MINIMUM_POLLING_FREQUENCY, 60, Integer.MAX_VALUE);
+                    new ZapNumberSpinner(BoastParam.MINIMUM_POLLING_FREQUENCY, 60, 6 * 60 * 60);
```
Comment (kingthorin) [inline addOns/oast/src/main/java/org/zaproxy/addon/oast/services/boast/BoastOptionsPanelTab.java:97]: I know it makes it a bit more verbose, but it's more obvious for maintenance etc.
```suggestion
                    new ZapNumberSpinner(BoastParam.MINIMUM_POLLING_FREQUENCY, TimeUnit.HOURS.toSeconds(6));
```

**Pull request merged**